common/stats/stats_test.go

```diff
@@ -109,52 +107,50 @@ func TestNonLatching(t *testing.T) {
-	statIface, cancelFn := NewLatchedStatsReceiver(time.Hour)
-	stat := statIface.(*defaultStatsReceiver)
-	captured := stat.capture()
-	cancelFn()
-
-	// Replace latchCh and construct channel to step through latch()
-	stat.latchCh = make(chan chan StatsRegistry)
 	ct := make(chan time.Time, 2)
-	respCh := make(chan bool)
 	Time = NewTestTime(time.Unix(0, 0), 0, ct)
-	ctx, cancel := context.WithCancel(context.Background())
-
-	// Don't capture (ticker=0, firstSnapshotAt=1s)
-	go latch(stat, captured, stat.latchCh, Time.NewTicker(0), Time.Now().Add(time.Second), ctx, respCh)
+	statIface, cancelFn := NewLatchedStatsReceiver(time.Second)
+	stat := statIface.(*defaultStatsReceiver)
+	defer cancelFn()
 
-	// Registry should not be captured in latch() until we accrue measurements.
+	// Captured registry should initially be empty even though we added a counter.
 	stat.Counter("counter")
-	ct <- Time.Now()
-	<-respCh
 	rendered := string(stat.Render(true))
 	if rendered != "{}" {
 		t.Fatal("Expected empty latch with time=0: ", rendered)
 	}
 
-	// Replace latchCh and construct channel to step through latch()
-	cancel()
-	stat.latchCh = make(chan chan StatsRegistry)
-	ct = make(chan time.Time, 2)
-	Time = NewTestTime(time.Unix(0, 0), 0, ct)
-	ctx, cancel = context.WithCancel(context.Background())
-
-	// Captures immediately. (ticker=1m, firstSnapshotAt=0)
-	go latch(stat, captured, stat.latchCh, Time.NewTicker(0), Time.Now(), ctx, respCh)
+	// Captured registry should still be empty.
+	ensureChannelSend(cap(ct), func() { ct <- Time.Now().Add(1) })
+	rendered = string(stat.Render(true))
+	if rendered != "{}" {
+		t.Fatal("Expected empty latch with time=1: ", rendered)
+	}
 
-	// Captured registry should be updated here and render should pick that up.
-	ct <- Time.Now().Add(time.Minute)
-	<-respCh
+	// Should be captured after enough time has passed.
+	ensureChannelSend(cap(ct), func() { ct <- Time.Now().Add(time.Minute) })
 	rendered = string(stat.Render(true))
 	if rendered == "{}" {
 		t.Fatal("Expected non-empty latch with time=0: ", rendered)
 	}
-	if stat.Counter().Count() != 0 {
-		t.Fatal("Expected counter to be cleared after non-latch")
+}
+
+// Three sends are required to ensure the first is handled, not just goroutine-queued...
+// This function may seem hacky, but it's the simplest way to do ordered testing when we're
+// select'ing on two+ channels: one channel to trigger updates, another accepting
+// requests and immediately responding with a cached update.
+//
+// The Memory Model is light on details regarding select'ing from multiple channels,
+// but testing shows that the following sequence (or something similar) will happen
+// for unbuffered channels:
+// - first send is received/popped but doesn't enter the case statement. The channel is ready again.
+// - second send is is received and blocks further sends as the first send is still pending.
+// - third send blocks until the receive case statement for the first send is fully executed.
```

Comment:
Yes, that would be the expected outcome. The relationship between goroutines and channels is more nuanced though and I can't find firm proof that the handoff is synchronous. Is it self evident that it does the right thing? I can always go to the source to mollify myself (I probably should have done that already.)





On Fri, Jul 15, 2016 at 10:18 AM -0700, "Dan Bentley" <notifications@github.com> wrote:












In common/stats/stats_test.go:
>  	rendered = string(stat.Render(true))
>  	if rendered == "{}" {
>  		t.Fatal("Expected non-empty latch with time=0: ", rendered)
>  	}
> -	if stat.Counter().Count() != 0 {
> -		t.Fatal("Expected counter to be cleared after non-latch")
> +}
> +
> +// Three sends are required to ensure the first is handled, not just goroutine-queued...
> +// This function may seem hacky, but it's the simplest way to do ordered testing when we're
> +// select'ing on two+ channels: one channel to trigger updates, another accepting
> +// requests and immediately responding with a cached update.
> +//
> +// The Memory Model is light on details regarding select'ing from multiple channels,
> +// but testing shows that the following sequence (or something similar) will happen
> +// for unbuffered channels:
> +// - first send is received/popped but doesn't enter the case statement. The channel is ready again.
> +// - second send is is received and blocks further sends as the first send is still pending.
> +// - third send blocks until the receive case statement for the first send is fully executed.



If a send to an unbuffered channel succeeds, then the receive on the other side succeeded, which means it has entered the case and will not try another receive until the next iteration through the loop... So once should be enough? Or am I missing something?



—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub, or mute the thread.